drivers/comparator/comparator_nrf_comp.c

```diff
@@ -0,0 +1,771 @@
+#endif
+	.refsel = SHIM_NRF_COMP_DT_INST_REFSEL(0),
+	.th_down = SHIM_NRF_COMP_DT_INST_TH_DOWN(0),
+	.th_up = SHIM_NRF_COMP_DT_INST_TH_UP(0),
+};
+#else
+static const struct comp_nrf_comp_diff_config shim_nrf_comp_config0 = {
+	.psel = SHIM_NRF_COMP_DT_INST_PSEL(0),
+	.sp_mode = SHIM_NRF_COMP_DT_INST_SP_MODE(0),
+	.isource = SHIM_NRF_COMP_DT_INST_ISOURCE(0),
+	.extrefsel = SHIM_NRF_COMP_DT_INST_EXTREFSEL(0),
+	.enable_hyst = SHIM_NRF_COMP_DT_INST_ENABLE_HYST(0),
+};
+#endif
+
+static struct shim_nrf_comp_data shim_nrf_comp_data0;
+
+#if CONFIG_PM_DEVICE
+static bool shim_nrf_comp_is_resumed(void)
+{
+	enum pm_device_state state;
+
+	(void)pm_device_state_get(DEVICE_DT_INST_GET(0), &state);
+	return state == PM_DEVICE_STATE_ACTIVE;
+}
+#else
+static bool shim_nrf_comp_is_resumed(void)
+{
+	return true;
+}
+#endif
+
+static void shim_nrf_comp_start(void)
+{
+	if (shim_nrf_comp_data0.started) {
+		return;
+	}
+
+	nrfx_comp_start(shim_nrf_comp_data0.event_mask, 0);
+	shim_nrf_comp_data0.started = true;
+}
+
+static void shim_nrf_comp_stop(void)
+{
+	if (!shim_nrf_comp_data0.started) {
+		return;
+	}
+
+	nrfx_comp_stop();
+	shim_nrf_comp_data0.started = false;
+}
+
+static int shim_nrf_comp_pm_callback(const struct device *dev, enum pm_device_action action)
+{
+	ARG_UNUSED(dev);
+
+	if (action == PM_DEVICE_ACTION_RESUME) {
+		shim_nrf_comp_start();
+	}
+
+#if CONFIG_PM_DEVICE
+	if (action == PM_DEVICE_ACTION_SUSPEND) {
+		shim_nrf_comp_stop();
+	}
+#endif
+
+	return 0;
+}
+
+#if (NRF_COMP_HAS_AIN_AS_PIN)
+static int shim_nrf_comp_psel_to_nrf(enum comp_nrf_comp_psel shim,
+				     nrf_comp_input_t *nrf)
+{
+	if (shim >= ARRAY_SIZE(shim_nrf_comp_ain_map)) {
+		return -EINVAL;
+	}
+
+	*nrf = shim_nrf_comp_ain_map[(uint32_t)shim];
+	return 0;
+}
+#else
+static int shim_nrf_comp_psel_to_nrf(enum comp_nrf_comp_psel shim,
+				     nrf_comp_input_t *nrf)
+{
+	int ret = 0;
+
+	switch (shim) {
+	case COMP_NRF_COMP_PSEL_AIN0:
+		*nrf = NRF_COMP_INPUT_0;
+		break;
+
+	case COMP_NRF_COMP_PSEL_AIN1:
+		*nrf = NRF_COMP_INPUT_1;
+		break;
+
+	case COMP_NRF_COMP_PSEL_AIN2:
+		*nrf = NRF_COMP_INPUT_2;
+		break;
+
+	case COMP_NRF_COMP_PSEL_AIN3:
+		*nrf = NRF_COMP_INPUT_3;
+		break;
+
+#if defined(COMP_PSEL_PSEL_AnalogInput4)
+	case COMP_NRF_COMP_PSEL_AIN4:
+		*nrf = NRF_COMP_INPUT_4;
+		break;
+#endif
+
+#if defined(COMP_PSEL_PSEL_AnalogInput5)
+	case COMP_NRF_COMP_PSEL_AIN5:
+		*nrf = NRF_COMP_INPUT_5;
+		break;
+#endif
+
+#if defined(COMP_PSEL_PSEL_AnalogInput6)
+	case COMP_NRF_COMP_PSEL_AIN6:
+		*nrf = NRF_COMP_INPUT_6;
+		break;
+#endif
+
+#if defined(COMP_PSEL_PSEL_AnalogInput7)
+	case COMP_NRF_COMP_PSEL_AIN7:
+		*nrf = NRF_COMP_INPUT_7;
+		break;
+#endif
+
+#if defined(COMP_PSEL_PSEL_VddDiv2)
+	case COMP_NRF_COMP_PSEL_VDD_DIV2:
+		*nrf = NRF_COMP_VDD_DIV2;
+		break;
+#endif
+
+#if defined(COMP_PSEL_PSEL_VddhDiv5)
+	case COMP_NRF_COMP_PSEL_VDDH_DIV5:
+		*nrf = NRF_COMP_VDDH_DIV5;
+		break;
+#endif
+
+	default:
+		ret = -EINVAL;
```

Comment:
how about `return -EINVAL` here, `return 0` at the end and drop the ret variable? one less thing to track, I think it makes these a bit easier to follow, same below, you actually do it already in some of the functions anyway